File: TradeItIosEmsApi/TradeItTradeService.m
//
//  TradeItTradeService.m
//  TradeItIosEmsApi
//
//  Created by Antonio Reyes on 1/15/16.
//  Copyright © 2016 TradeIt. All rights reserved.
//

#import "TradeItTradeService.h"
#import "TradeItRequestResultFactory.h"
#import "TradeItPreviewTradeResult.h"
#import "TradeItPlaceTradeResult.h"

@implementation TradeItTradeService

- (id)initWithSession:(TradeItSession *) session {
    self = [super init];
    if (self) {
        self.session = session;
    }
    return self;
}

- (void)previewTrade:(TradeItPreviewTradeRequest *)order withCompletionBlock:(void (^)(TradeItResult *)) completionBlock {
    order.token = self.session.token;
    
    NSURLRequest * request = [TradeItRequestResultFactory buildJsonRequestForModel:order
                                                                         emsAction:@"order/previewStockOrEtfOrder"
                                                                       environment:self.session.connector.environment];

    [self.session.connector sendEMSRequest:request
                            forResultClass:[TradeItPreviewTradeResult class]
                       withCompletionBlock:completionBlock];
}

- (void)placeTrade:(TradeItPlaceTradeRequest *)order withCompletionBlock:(void (^)(TradeItResult *))completionBlock {
    order.token = self.session.token;
    
    NSURLRequest *request = [TradeItRequestResultFactory buildJsonRequestForModel:order
                                                                        emsAction:@"order/placeStockOrEtfOrder"
                                                                      environment:self.session.connector.environment];

    [self.session.connector sendEMSRequest:request
                            forResultClass:[TradeItPlaceTradeResult class]
                       withCompletionBlock:completionBlock];
}

@end

File: TradeItIosTicketSDK2/TradeItTradeService.swift
@objc public class TradeItTradeService: NSObject {
    private let session: TradeItSession

    init(session: TradeItSession) {
        self.session = session
    }

    func previewTrade(
        _ data: TradeItPreviewTradeRequest,
        onSuccess: @escaping (TradeItPreviewTradeResult) -> Void,
        onFailure: @escaping (TradeItErrorResult) -> Void
    ) {
        data.token = self.session.token

        let request = TradeItRequestResultFactory.buildJsonRequest(
            for: data,
            emsAction: "order/previewStockOrEtfOrder",
            environment: self.session.connector.environment
        )

        self.session.connector.sendEMSRequest(request, forResultClass: TradeItPreviewTradeResult.self) { result in
            switch (result) {
            case let result as TradeItPreviewTradeResult: onSuccess(result)
            case let error as TradeItErrorResult: onFailure(error)
            default:
                onFailure(TradeItErrorResult(
                    title: "Preview failed",
                    message: "There was a problem previewing your order. Please try again."
                ))
            }
        }
    }

    func placeTrade(
        _ data: TradeItPlaceTradeRequest,
        onSuccess: @escaping (TradeItPlaceTradeResult) -> Void,
        onFailure: @escaping (TradeItErrorResult) -> Void
    ) {
        data.token = self.session.token

        let request = TradeItRequestResultFactory.buildJsonRequest(
            for: data,
            emsAction: "order/placeStockOrEtfOrder",
            environment: self.session.connector.environment
        )

        self.session.connector.sendEMSRequest(request, forResultClass: TradeItPlaceTradeResult.self) { result in
            switch (result) {
            case let result as TradeItPlaceTradeResult: onSuccess(result)
            case let error as TradeItErrorResult: onFailure(error)
            default:
                onFailure(TradeItErrorResult(
                    title: "Invalid response sent from the server",
                    message: "Please check your active orders and try again."
                ))
            }
        }
    }
}
